Error in ManagedGit when dealing with non-packed lightweight tags

I am trying to build https://github.com/sshnet/SSH.NET/pull/1299 in my local fork. The repo has some tags which are lightweight tags e.g. 2023.0.0. The build fails locally with

```
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly.
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: Nerdbank.GitVersioning.GitException: Got a com instead of a tag when opening object 1c7166a002d7633fe1595b3df4634ba0ef6e3138
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.TryGetObjectByPath(GitObjectId sha, String objectType, Stream& value)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.TryGetObjectBySha(GitObjectId sha, String objectType, Stream& value)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.<LookupTags>g__HandleCandidate|44_0(GitObjectId pointsAt, String tagName, Boolean isPeeled, <>c__DisplayClass44_0& )
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.LookupTags(GitObjectId objectId)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.ManagedGitContext.get_HeadTags()
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset)
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner() in D:\a\1\s\src\Nerdbank.GitVersioning.Tasks\GetBuildVersion.cs:line 240
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
1>C:\Users\rhague\.nuget\packages\nerdbank.gitversioning\3.7.48-alpha\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

The problem appears to be that ManagedGit is assuming the object is of type "tag" (annotated) whereas it is of type "commit" (lightweight). The error is here:

https://github.com/dotnet/Nerdbank.GitVersioning/blob/549f6f4ed81ac37dc7be2968690fec9904e87e08/src/NerdBank.GitVersioning/ManagedGit/GitRepository.cs#L782

and it comes from

https://github.com/dotnet/Nerdbank.GitVersioning/blob/549f6f4ed81ac37dc7be2968690fec9904e87e08/src/NerdBank.GitVersioning/ManagedGit/GitRepository.cs#L657

The reproduction is a little convoluted because there is no error on a fresh clone with .pack files, and it also requires the tags to be "unpacked" (in .git/refs/tags rather than in .git/packed-refs). That's because of the `!isPeeled` check in the above line[^1].

```bash
git clone https://github.com/sshnet/SSH.NET.git
cd SSH.NET
git fetch origin pull/1299/head:repro
git checkout 5ab40e # or otherwise git checkout repro
git show-ref 2023.0.0 # 1c7166a002d7633fe1595b3df4634ba0ef6e3138 refs/tags/2023.0.0
git cat-file -t 2023.0.0 # commit
grep -r 2023.0.0 .git/ # .git/packed-refs:1c7166a002d7633fe1595b3df4634ba0ef6e3138 refs/tags/2023.0.0
dotnet build # this works
git tag | xargs git tag -d && git fetch --tags # "unpack" the tags (now 2023.0.0 is a file in refs/tags/)
mv .git/objects/pack/*.pack .
for fn in ./*.pack; do git unpack-objects < "$fn"; done # unpack objects
dotnet build # this fails
```

Arguably these tags *should* be annotated, but supposing it makes sense to consider lightweight tags in this process, perhaps `TryGetObjectByPath` should be returning false rather than throwing (like `GitPack.TryGetObject` effectively seems to do), and then things should work correctly? I don't know how this logic fits into the wider process.

[^1]: `isPeeled` comes from the `out` parameter of `EnumeratePackedRefsRaw` and relates to the packed-refs file as a whole rather than whether a particular object is peeled. That doesn't seem intuitively right but then I don't know anything about peeling.

--- a/src/NerdBank.GitVersioning/ManagedGit/GitRepository.cs
+++ b/src/NerdBank.GitVersioning/ManagedGit/GitRepository.cs
@@ -779,7 +779,8 @@ public class GitRepository : IDisposable
 
         if (string.CompareOrdinal(objectStream.ObjectType, objectType) != 0)
         {
-            throw new GitException($"Got a {objectStream.ObjectType} instead of a {objectType} when opening object {sha}");
+            value = null;
+            return false;
         }
 
         value = objectStream;

--- a/test/Nerdbank.GitVersioning.Tests/ManagedGit/GitRepositoryTests.cs
+++ b/test/Nerdbank.GitVersioning.Tests/ManagedGit/GitRepositoryTests.cs
@@ -272,6 +272,21 @@ public class GitRepositoryTests : RepoTestBase
     }
 
     [Fact]
+    public void TryGetObjectByShaAndWrongTypeTest()
+    {
+        this.InitializeSourceControl();
+        this.AddCommits(2);
+
+        var headObjectId = GitObjectId.Parse(this.LibGit2Repository.Head.Tip.Sha);
+
+        using (var repository = GitRepository.Create(this.RepoPath))
+        {
+            Assert.False(repository.TryGetObjectBySha(headObjectId, "tree", out Stream value));
+            Assert.Null(value);
+        }
+    }
+
+    [Fact]
     public void GetMissingObjectByShaTest()
     {
         this.InitializeSourceControl();

--- a/test/Nerdbank.GitVersioning.Tests/VersionOracleTests.cs
+++ b/test/Nerdbank.GitVersioning.Tests/VersionOracleTests.cs
@@ -1048,7 +1048,7 @@ public abstract class VersionOracleTests : RepoTestBase
         // Assert that we don't see any tags.
         Assert.Empty(oracle.Tags);
 
-        // Create a lighweight tag.
+        // Create a lightweight tag.
         this.LibGit2Repository.ApplyTag("mytag");
 
         // Refresh our context before asking again.

--- a/test/Nerdbank.GitVersioning.Tests/VersionOracleTests.cs
+++ b/test/Nerdbank.GitVersioning.Tests/VersionOracleTests.cs
@@ -1048,7 +1048,7 @@ public abstract class VersionOracleTests : RepoTestBase
         // Assert that we don't see any tags.
         Assert.Empty(oracle.Tags);
 
-        // Create a tag.
+        // Create a lighweight tag.
         this.LibGit2Repository.ApplyTag("mytag");
 
         // Refresh our context before asking again.
@@ -1057,6 +1057,16 @@ public abstract class VersionOracleTests : RepoTestBase
 
         // Assert that we see the tag.
         Assert.Equal("refs/tags/mytag", Assert.Single(oracle2.Tags));
+
+        // Add another commit
+        this.AddCommits(1);
+
+        // Refresh our context before asking again.
+        this.Context = this.CreateGitContext(this.RepoPath);
+        VersionOracle oracle3 = new(this.Context);
+
+        // Assert that HEAD is not pointing to the tag.
+        Assert.Empty(oracle3.Tags);
     }
 
     [Fact]
@@ -1070,7 +1080,7 @@ public abstract class VersionOracleTests : RepoTestBase
         // Assert that we don't see any tags.
         Assert.Empty(oracle.Tags);
 
-        // Create a tag.
+        // Create an annotated tag.
         this.LibGit2Repository.ApplyTag("mytag", this.Signer, "my tag");
 
         // Refresh our context before asking again.
@@ -1079,6 +1089,16 @@ public abstract class VersionOracleTests : RepoTestBase
 
         // Assert that we see the tag.
         Assert.Equal("refs/tags/mytag", Assert.Single(oracle2.Tags));
+
+        // Add another commit
+        this.AddCommits(1);
+
+        // Refresh our context before asking again.
+        this.Context = this.CreateGitContext(this.RepoPath);
+        VersionOracle oracle3 = new(this.Context);
+
+        // Assert that HEAD is not pointing to the tag.
+        Assert.Empty(oracle3.Tags);
     }
 
     [Fact]

